src/Lib.P2P/PeerConnection.cs

```diff
@@ -0,0 +1,415 @@
+        public TaskCompletionSource<Peer> IdentityEstablished { get; } = new TaskCompletionSource<Peer>();
+
+        /// <summary>
+        ///   When the connection was last used.
+        /// </summary>
+        public DateTime LastUsed => statsStream.LastUsed;
+
+        /// <summary>
+        ///   Number of bytes read over the connection.
+        /// </summary>
+        public long BytesRead => statsStream.BytesRead;
+
+        /// <summary>
+        ///   Number of bytes written over the connection.
+        /// </summary>
+        public long BytesWritten => statsStream.BytesWritten;
+
+        /// <summary>
+        ///  Establish the connection with the remote node.
+        /// </summary>
+        /// <param name="securityProtocols"></param>
+        /// <param name="cancel"></param>
+        /// <remarks>
+        ///   This should be called when the local peer wants a connection with
+        ///   the remote peer.
+        /// </remarks>
+        public async Task InitiateAsync(IEnumerable<IEncryptionProtocol> securityProtocols,
+            CancellationToken cancel = default(CancellationToken))
+        {
+            await EstablishProtocolAsync("/multistream/", cancel).ConfigureAwait(false);
+
+            // Find the first security protocol that is also supported by the remote.
+            var exceptions = new List<Exception>();
+            foreach (var protocol in securityProtocols)
+            {
+                try
+                {
+                    await EstablishProtocolAsync(protocol.ToString(), cancel).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    exceptions.Add(e);
+                    continue;
+                }
+
+                await protocol.EncryptAsync(this, cancel).ConfigureAwait(false);
+                break;
+            }
+
+            if (!SecurityEstablished.Task.IsCompleted)
+                throw new AggregateException("Could not establish a secure connection.", exceptions);
+
+            await EstablishProtocolAsync("/multistream/", cancel).ConfigureAwait(false);
+            await EstablishProtocolAsync("/mplex/", cancel).ConfigureAwait(false);
+
+            var muxer = new Muxer
+            {
+                Channel = Stream,
+                Initiator = true,
+                Connection = this
+            };
+            muxer.SubstreamCreated += (s, e) => _ = ReadMessagesAsync(e, CancellationToken.None);
+            MuxerEstablished.SetResult(muxer);
+
+            _ = muxer.ProcessRequestsAsync();
+        }
+
+        /// <summary>
+        ///   TODO:
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="cancel"></param>
+        /// <returns></returns>
+        public Task EstablishProtocolAsync(string name, CancellationToken cancel)
+        {
+            return EstablishProtocolAsync(name, Stream, cancel);
+        }
+
+        /// <summary>
+        ///   TODO:
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)